test/test_pghoard.py

```diff
@@ -744,17 +745,76 @@ def test_startup_walk_for_missed_uncompressed_file_type(self, file_type: FileTyp
-        compressed_wal_path, _ = self.pghoard.create_backup_site_paths(self.test_site)
-        with open(os.path.join(compressed_wal_path, file_name), "wb") as fp:
+        backup_site_paths = self.pghoard.create_backup_site_paths(self.test_site)
+
+        if file_type is FileType.Wal:
+            compressed_file_path = backup_site_paths.compressed_xlog_path
+        else:
+            compressed_file_path = backup_site_paths.compressed_timeline_path
+
+        with open(os.path.join(compressed_file_path, file_name), "wb") as fp:
             fp.write(b"foo")
-        with open(os.path.join(compressed_wal_path, f"{file_name}.metadata"), "wb") as fp:
+        with open(os.path.join(compressed_file_path, f"{file_name}.metadata"), "wb") as fp:
             fp.write(b"{}")
         self.pghoard.startup_walk_for_missed_files()
         assert self.pghoard.compression_queue.qsize() == 0
         assert self.pghoard.transfer_queue.qsize() == 1
         upload_event = self.pghoard.transfer_queue.get(timeout=1.0)
         assert upload_event.file_type == file_type
 
+    @pytest.mark.parametrize(
+        "file_type, file_name, invalid_compressed_file_name", [
+            (FileType.Wal, "000000010000000000000005", "000000010000000000000006"),
+            (FileType.Timeline, "00000003.history", "00000004.history"),
+        ]
+    )
+    def test_startup_walk_skip_compression_if_already_compressed(
+        self,
+        file_type: FileType,
+        file_name: str,
+        invalid_compressed_file_name: str,
+    ) -> None:
+        """
+        Tests the scenario where an uncompressed file was already compressed, but was not deleted.
+        """
+        backup_site_paths = self.pghoard.create_backup_site_paths(self.test_site)
+        uncompressed_wal_path = backup_site_paths.uncompressed_files_path
+        compressed_file_path = (
+            backup_site_paths.compressed_timeline_path
+            if file_type is FileType.Timeline else backup_site_paths.compressed_xlog_path
+        )
+
+        # generate  uncompressed/compressed files
+        with open(os.path.join(uncompressed_wal_path, file_name), "wb") as fp:
+            fp.write(b"random")
+
+        with open(os.path.join(uncompressed_wal_path, invalid_compressed_file_name), "wb") as fp:
+            fp.write(b"random")
+
+        # compressed
+        with open(os.path.join(compressed_file_path, file_name), "wb") as fp:
+            fp.write(b"random")
+
+        with open(os.path.join(compressed_file_path, f"{file_name}.metadata"), "wb") as fp:
+            fp.write(b"{}")
+
+        # invalid compressed file should not have a metadata
+        with open(os.path.join(compressed_file_path, invalid_compressed_file_name), "wb") as fp:
+            fp.write(b"random")
+
+        self.pghoard.startup_walk_for_missed_files()
+
+        # only one file should be added for compression (invalid compressed one)
+        assert self.pghoard.compression_queue.qsize() == 1
```

Comment:
It would not hurt to add checks that the correct messages are added to the queue, as we don't know if the correct files were added. The same below.

Reply:
done here https://github.com/aiven/pghoard/pull/592/commits/60d6bf64a5801efe430f5b722bb53c9e8e330e7f